Add code for Ansible Galaxy release automation for Collections

ericsysmin opened the pull request:
##### SUMMARY
Add documentation for Ansible Galaxy release automation via GitHub Action Workflow

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

abadger commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,218 @@
+        copy:
+          content: |
+            token: {{ lookup('env','ANSIBLE_GALAXY_TOKEN') }}
+          dest: ~/.ansible/galaxy_token
```
If you write this to disk, make sure the file's nose is 0600

ericsysmin replied:
Good point will make the change

abadger commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,218 @@
+      - name: Run role test
+        run: >-
+          ansible-playbook -i 'localhost,' build/galaxy_deploy.yml -e "github_tag=${{ github.ref }}"
+
```
Is there a reason to have both a playbook and an action? Why not put everything into an action or create different instructions for people who want to do things differently

ericsysmin replied:
The playbook does the templating and processing of the values. What action do you see here?

abadger replied:
I mean, the GitHub action. Although templating might be less easy with a raw action.

ericsysmin replied:
Let's see where the discussion ends up, maybe that ends up being where we take this now that people want all these other steps in place.

webknjaz replied:
Yeah, it could be a great idea to actually create our own action. But I'd not mix build&publish, because one should inject some testing in between, so it'd be two actions.

ericsysmin replied:
What are the proposed steps then?

webknjaz replied:
If you're up to writing a few lines of bash and a dockerfile, you could create an action or two. The question is whether it's overkill for this task, though.

abadger commented:
One overarching question, is this still gplv3 compliant?  Is everything used to build the collection inside the collection tarball (i mean, the .Github directory and the deploy playbook)?

ericsysmin commented:
@abadger were not uploading the tar.gz created by GitHub, were uploading the asset made by Ansible Collection build command. But there is a tar.gz you could download just like you currently do on roles, or any release on GitHub. I just made it easier to tag the release the way you want. However currently Ansible build does include molecule, build, and other directories, not in excludes (supported in 2.10).

felixfontein commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,218 @@
+  └── templates
+      └── galaxy.yml.j2
+
+Instead of having the :code:`galaxy.yml` file in our root, we will be generating
```
One disadvantage of this approach is that tests which need to know the next planned version number cannot read it from anywhere. Such tests don't yet exist in ansible-test, but I'm working on some (ansible/ansible#68646, ansible/ansible#66920). This would then mean that the version number must be supplied some other way in CI, and that it needs to be stored somewhere else.

ericsysmin replied:
Couldn't we generate it then run ansible-test?

felixfontein replied:
How should ansible-test know the next released version of the collection if it can't read it from galaxy.yml?

ericsysmin replied:
@geerlingguy have you thought about this? I know you're also using this process on one of your collections at the moment but with Travis-CI.

geerlingguy replied:
Yeah... this is one of the annoying things that I don't know how we'd be able to solve it in a sane way. Basically you either have to tie your CI system back to git, so it can modify `galaxy.yml` then push back a commit—but that would happen _after_ you create the git tag.

Otherwise, for every release, the process becomes much more annoying (compared to roles, where you just run `git tag x.y.z && git push --tags` and that's it):

  1. Choose what you want your tag to be. Update `galaxy.yml` with that tag.
  2. Push that commit with the new tag in it... make sure CI passes.
  3. `git tag x.y.z && git push --tags`
  4. Wait for CI process to then build and publish the release.

Note that I've also opened [[Proposal] Support building and publishing collections server side](https://github.com/ansible/galaxy_ng/issues/133), which might also alleviate a lot of the annoyance over managing collections. Roles are/[[were](https://github.com/ansible/galaxy_ng/issues/58)?] much simpler to maintain.

felixfontein replied:
Another (related) question: where in that process do you update the changelog / release notes? That should also contain the new version number (and more release-dependent information).

ericsysmin replied:
When you create a release through GitHub. However, do we even have an official process for updating changelogs on collections yet?

felixfontein replied:
I don't think we have. It's a related problem though, since the changelog also needs to know the release version, and after the release needs to be updated in some way so that new entries can be added for the next version.

I'm currently working on generalizing Ansible's changelog generator to collections (https://github.com/felixfontein/ansible-changelog/); it also uses galaxy.yml to determine the relesae version by default (you can still specify it though).

At least with this changelog generator, the release + changelog process could look like this:

0. (Not really part of the process: galaxy.yml contains the version of the next release; changes come with changelog fragments.)
1. Run `ansible-changelog release`, which creates a new release in the changelog with the current set of fragments. (Also adds a list of new modules/plugins when `version_added` is correctly used.)
2. Commit updated `changelog.yaml` file and generated .rst version of changelog. (The `.yaml` file is both metadata and what is used by ACD to compile combined changelog.)
3. Create tag.
4. Build and upload collection.
5. Update `galaxy.yml` for the next version and commit.

(And somewhen also push.)

This could be automated (by a script, playbook, github action, a combination of these, ...), except of course determining the next release version, which has to be done manually in some way.

ericsysmin replied:
1. Run Ansible playbook to generate galaxy.yml
2. Run ansible-test
3. Run ansible-changelog
4. Run ansible collection build
5. Run ansible collection publish

However,  by the time you even do a release isn't it expected you already had a branch with the release you're releasing from that has been being tested? Just a thought.

felixfontein replied:
You might also want to run ansible-test *before* starting the release process :)

Also, you're missing a commit of the updated changelog in your steps.

> However, by the time you even do a release isn't it expected you already had a branch with the release you're releasing from that has been being tested? Just a thought.

That depends a lot on how the versioning/release process for the collection looks like. There will be different approaches, I guess, and right now that is also [under discussion](https://github.com/ansible-collections/overview/issues/37) - well, not that there is much of a discussion right now.

ericsysmin replied:
I mean who starts a release without ansible-test, molecule, etc running before you create a release. Each PR and Push should be getting tested before any of these steps. Because I think it's relative to this discussion, please checkout these documents and pages.

https://github.com/geerlingguy/ansible-collection-php_roles/blob/master/.travis.yml
https://github.com/geerlingguy/ansible-collection-k8s/blob/master/.travis.yml
https://www.jeffgeerling.com/blog/2020/automatically-building-and-publishing-ansible-galaxy-collections
https://www.jeffgeerling.com/blog/2019/how-add-integration-tests-ansible-collection-ansible-test
https://ericsysmin.com/2020/05/02/ansible-collections-automating-the-release-process-to-galaxy/
https://ericsysmin.com/2020/04/30/ansible-collections-role-tests-with-molecule/

gundalow replied:
@pabelanger how are you updating the version in galaxy.yml in Zuul?

felixfontein replied:
@ericsysmin I mean running sanity tests which actually need to know the version number, like checking for deprecated features of modules, or making sure that PRs contain the correct value of `version_added`. That is of course not relevant for collections with roles in them (from the links I see that you seem to be more interested in these); for collections with modules, especially large and diverse ones like community.general and community.network, these tests are more important (IMO very important - for the same reasons that these tests existed and still exist for PRs in ansible/ansible itself).

pabelanger replied:
@gundalow https://github.com/ansible-network/releases/blob/master/ansible_releases/cmd/generate_ansible_collection.py

we use pbr, to generate semver based on git commit. Then repopulate galaxy.yml with the version number

ericsysmin replied:
@felixfontein I don't have many collections where I develop modules themselves rather work on other's modules or adding them for other projects, so I don't have a large collection to test or use. So it's relevant for the purpose of this commit, and this is why we're collaborating on this. If you can make some commits to this PR that would be great.

abadger commented:
> @abadger were not uploading the tar.gz created by GitHub, were uploading the asset made by Ansible Collection build command. But there is a tar.gz you could download just like you currently do on roles, or any release on GitHub. I just made it easier to tag the release the way you want. However currently Ansible build does include molecule, build, and other directories, not in excludes (supported in 2.10).

<Nod>. Yeah, we have to be careful not to exclude things used to build the asset.  Otherwise it no longer counts as source for the gpl.

ericsysmin commented:
> > @abadger were not uploading the tar.gz created by GitHub, were uploading the asset made by Ansible Collection build command. But there is a tar.gz you could download just like you currently do on roles, or any release on GitHub. I just made it easier to tag the release the way you want. However currently Ansible build does include molecule, build, and other directories, not in excludes (supported in 2.10).
> 
> . Yeah, we have to be careful not to exclude things used to build the asset. Otherwise it no longer counts as source for the gpl.

Ah, then that's a problem for Ansible Collections in general as they do exclude directories by default.

abadger commented:
Yep :-(. I want to make sure we do it right but I'm not sure what to do about other people yet.  (It's a problem that attends from Galaxy only hosting the built collections, not a source tarball)

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,218 @@
+  release:
+    runs-on: ubuntu-18.04
+    env:
+      ANSIBLE_GALAXY_TOKEN: ${{ secrets.ANSIBLE_GALAXY_TOKEN }}
```
I'd rather open this token only to the upload step for security reasons

webknjaz replied:
Even better if you could pass it via the args of the ansible-playbook CLI command (all the secrets are automatically sanitized in GitHub Workflows output)

ericsysmin replied:
make the code suggestion and we can commit it

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,218 @@
+      - created
+  jobs:
+  release:
+    runs-on: ubuntu-18.04
```
Probably better to use
```suggestion
    runs-on: ubuntu-latest
```

ericsysmin replied:
So what do you do when 20.04 is the latest and doesn't have the expected packages or support?

ericsysmin replied:
Assuming latest is best is never a good way to go, at least from any operations perspective because latest is unpredictable in most including docker as a tag of latest can and may get moved to a release your code isn't tested on.

webknjaz replied:
there was a problem when GH was moving fast and dropped support for previous macOS versions, for example. and it was PITA to upgrade the hardcoded thing everywhere. I'm just trying to avoid the same. Besides, it's not like you're using some extraordinary software here that may suddenly disappear.

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,218 @@
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install ansible
```
It's inconsistent with the previous command. And probably it's better to install under the user dir:
```suggestion
          python -m pip install ansible --user
          echo "::add-path::${{ env.HOME }}/.local/bin"
```

webknjaz replied:
Also, have you checked that ansible that happens to be already pre-installed on the runners is good enough for building&publishing collections? It'd mean that it may be unnecessary to also install it manually.

ericsysmin replied:
Didn't know about the `::add-path::` value.

ericsysmin replied:
Should we modify the pip upgrade step as well to use --user then?

webknjaz replied:
It'd install pip into a user dir too. Better question would be — do we really have to upgrade pip?

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+        file:
+          path: ~/.ansible
+          state: directory
+      - name: Write the Galaxy token to ~/.ansible/galaxy_token
```
Why is it needed to write this on disk at all?

ericsysmin replied:
Because it's not passed as a param to the publish, instead we're reading from .ansible

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+      - name: Build the collection. # noqa 503
+        command: >
+          ansible-galaxy collection build
+          chdir=../
```
this chdir thing looks weird. did you mean something like
```suggestion
        args:
          chdir: ../
```

ericsysmin replied:
No, what I pasted actually works, @geerlingguy used it and I used it that same way

webknjaz replied:
I believe that it works. But my concern here is readability.

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+
+      - name: Run the Ansible Galaxy deploy playbook
+        run: >-
+          ansible-playbook -i 'localhost,' build/galaxy_deploy.yml -e "github_tag=${{ github.ref }}"
```
I think I'd rather pass the secret like this:
```suggestion
          ansible-playbook -i 'localhost,' build/galaxy_deploy.yml -e "github_tag=${{ github.ref }}" -e "ansible_galaxy_token=${{ secrets.ANSIBLE_GALAXY_TOKEN }}"
```

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+parsed from GitHub as it passes this value on release to the Workflow
+automatically.
+
+Configure ANSIBLE_GALAXY_TOKEN Secret
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
Configure ``ANSIBLE_GALAXY_TOKEN`` Secret
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+  release:
+    types:
+      - created
+  jobs:
+  release:
```
the indentation here is broken

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+    types:
+      - created
+  jobs:
+  release:
```
Also you should add a name for the job

ericsysmin replied:
it is named, however, tabs were messed up here, fixing it

webknjaz replied:
No, you only named a yaml key, but there's also `name: ` inside that can be specified for better representation in the UI.

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+
+.. code-block:: yaml
+
+  name: "release"
```
be consistent with the other strings in this snippet
```suggestion
  name: release
```

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,219 @@
+  name: collection_name
+
+  # The version of the collection. Must be compatible with semantic versioning
+  version: "{{ tag }}"
```
AFAICS this is the only value that's being rendered in this template, it seems to be overengineered and could be solved with a simple `sed`.

ericsysmin replied:
Awww, it's like we don't even need Ansible for things ;)

ericsysmin replied:
But if you notice tag isn't provided by the script, it's actually being processed as github.ref includes `/` inside because it's a release ref, not a number.

webknjaz replied:
yep, I saw that. Btw you could easily get it from the Git checkout like `setuptools-scm` does this which is even better because this'd be portable and could run from the local machine as well

ericsysmin replied:
I was just using the params that github provides internally, are you able to use github actions locally? or should we just delete this PR and move to using tags and using more than github workflow for releases?

webknjaz replied:
I said above assuming you're adding this extra ansible playbook layer because you want this automation to be portable and run both locally (w/o  GH actions) and in CI/CD.
If you only intend to have a GHA based solution you don't really need this many levels of indirection.

webknjaz commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,212 @@
+    types:
+      - created
+  jobs:
+    release:
```
```suggestion
    release:
      name: Build and release a collection
```

geerlingguy commented:
FYI: https://github.com/artis3n/ansible_galaxy_collection

It's a GitHub Action that has support for publishing collections. My main concern with any automation, however, is that right now for any collection where I upload my personal access token, if there's any way someone could sneak in a commit, they would immediately have access to publishing to _any_ other namespace/collection to which I have access—that needs to be an extremely bold and highlighted warning anywhere automation is concerned. See https://github.com/ansible/galaxy/issues/2275 and https://github.com/ansible/galaxy/issues/2276.

ericsysmin commented:
@geerlingguy I'm curious if using the token in the command to ansible exposes it on the github action, at least with the ENV var it doesn't expose it to the console I don't believe

ericsysmin commented on docs/galaxy_releases.rst:
```diff
@@ -0,0 +1,213 @@
+      - name: Publish the collection. # noqa 503
+        command: >
+          ansible-galaxy collection publish ./namespace-collection-{{ tag }}.tar.gz
+            --api-key={{ ansible_galaxy_token }}
```
We apparently can't do this because it would put the token in the std out
```n role test48s

Run ansible-playbook -i 'localhost,' build/galaxy_deploy.yml -e "github_tag=refs/tags/1.0.0"

PLAY [localhost] ***************************************************************

TASK [Ensure the ANSIBLE_GALAXY_TOKEN environment variable is set.] ************
skipping: [localhost]

TASK [Ensure the ~/.ansible directory exists.] *********************************
[DEPRECATION WARNING]: Distribution Ubuntu 18.04 on host localhost should use 
/usr/bin/python3, but is using /usr/bin/python for backward compatibility with 
ok: [localhost]```

ericsysmin replied:
Let me try and see what happens.

ericsysmin commented:
This is the tested output so far, and removing the apikey from the environment.

```

PLAY [localhost] ***************************************************************

TASK [Ensure that the ansible_galaxy_apikey exists] ****************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Template out the galaxy.yml file.] ***************************************
changed: [localhost]
[DEPRECATION WARNING]: Distribution Ubuntu 18.04 on host localhost should use 
/usr/bin/python3, but is using /usr/bin/python for backward compatibility with 
prior Ansible releases. A future Ansible release will default to using the 
discovered platform python for this host. See https://docs.ansible.com/ansible/
2.9/reference_appendices/interpreter_discovery.html for more information. This 
feature will be removed in version 2.12. Deprecation warnings can be disabled 
by setting deprecation_warnings=False in ansible.cfg.

TASK [Build the collection.] ***************************************************
changed: [localhost]

TASK [Publish the collection.] *************************************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=4    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

artis3n commented:
> @geerlingguy I'm curious if using the token in the command to ansible exposes it on the github action, at least with the ENV var it doesn't expose it to the console I don't believe

If you store the token in a repo secret, the output of the value is masked anywhere it is present in the logs: https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#about-encrypted-secrets

The risk with the token, as @geerlingguy points out, is that anyone with write access to the repo could push changes to the workflow file and have access to the secret to use in arbitrary bash. Forks do not get access to secrets in the original's repo, but then the workflow does not run successfully from a fork.

artis3n commented:
FWIW, once something like https://github.com/ansible/galaxy/issues/2276 exists I'd be happy to make supported changes to my Collection action that is mentioned by geerlingguy above.

ericsysmin commented:
> > @geerlingguy I'm curious if using the token in the command to ansible exposes it on the github action, at least with the ENV var it doesn't expose it to the console I don't believe
> 
> If you store the token in a repo secret, the output of the value is masked anywhere it is present in the logs: https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets#about-encrypted-secrets
> 
> The risk with the token, as @geerlingguy points out, is that anyone with write access to the repo could push changes to the workflow file and have access to the secret to use in arbitrary bash. Forks do not get access to secrets in the original's repo, but then the workflow does not run successfully from a fork.

That user would also need permissions to create the release as well.

ericsysmin commented:
That sounds almost like a fundamental flaw if you can't trust the people who manage the repo. They should also likely have access to the secret in the Settings -> Secrets right? Since the secret isn't available outside of the repo that also prevents anyone from getting that secret locally. No one should run this workflow locally. It's an instruction specifically for GitHub Actions Workflow which isn't local. However, another section could be dedicated to automating local release actions.

artis3n commented:
> They should also likely have access to the secret in the Settings -> Secrets right?

Repo owners with "owner" or "admin" privileges can add or edit the values of secrets, but you can't read the secret values once saved. "Write" contributors could not get to that menu.

> manage the repo. They should also likely have access to the secret in the Settings -> Secrets right? Since the secret isn't available outside of the repo that also prevents anyone from getting that secret locally

True, but if someone has write access they can add a `run` task on the workflow to curl the secret to their own server, or something. They could modify the `on` trigger so the workflow runs on their commit, no PR or release required.

> GitHub doesn't provide any way of preventing the exposure on secrets if the person has write access. 

Yup. Presumably someone with write access to a repo is trusted. I do think geerlingguy's point:

> if there's any way someone could sneak in a commit, they would immediately have access to publishing to any other namespace/collection to which I have access

is important to keep in mind. But, for my collection action (https://github.com/artis3n/ansible_galaxy_collection), I do have users pass their galaxy token via a Secret in their repo. That is the 'right' way to do it, but there is a risk with users with write permission inherent to Workflows that should probably be highlighted if a workflow suggestion is added to the docs.

> That sounds almost like a fundamental flaw if you can't trust the people who manage the repo.

Yeah, totally agree.

ericsysmin commented:
Ok, so maybe we should add these as details somewhere in the docs?

geerlingguy commented:
> That sounds almost like a fundamental flaw if you can't trust the people who manage the repo.

Just being the devil's advocate, when it comes to the possibility of any of the dozens of people maintaining the `community` collections which are going to be used across most Ansible installations after 2.10 having any of their personal tokens compromised across any of their repositories (if we have them all use their single personal access token for Galaxy across them all)... I'd rather have a belt-and-suspenders approach. One token that has access to everything in Galaxy is a bit risky.

This is probably not the main place for the security discussion, but we should make sure it is highlighted that if anyone were to get access to your token, they could post to any namespace/collection in Galaxy that you maintain.

gundalow commented:
For information, and `community.` and `ansible.` collections (which are hosted in https://github.com/ansible-collections/) will use Zuul to publish to Galaxy. 

This means even someone has admin rights on the repo they will not be able to get the Galaxy Token.

ericsysmin commented:
@gundalow is there a doc on how Zuul is used to publish to Galaxy?

pabelanger commented:
@ericsysmin we have a release-ansible-collection-galaxy job, that is available to all project which are using zuul for CI. https://dashboard.zuul.ansible.com/t/ansible/build/f3b10c3ed59c4584a081e245a5900dba/console is an example job run. Much like this code, there is a service account on galaxy, which we use the token for, to do the publish.

Then, when ever a project tags a release (or merges a commit) we build and publish to galaxy.

webknjaz commented:
@ericsysmin FYI if you want to explicitly hide something from the log, like secrets, you can also issue a workflow command: https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#masking-a-value-in-log

webknjaz commented:
> FYI there's already a promising GitHub Action for publishing collections by @artis3n here: https://github.com/artis3n/ansible_galaxy_collection

That action also mixes up two stages that are meant to be separate: build + publish. I'm strongly against promoting such an approach.
Ideally, it should be possible to test what's going to be published rather than just source. When build+publish are squashed there's a possibility that you'll be testing not exactly what the user will get. And this flow looks as follows:
1. test the src.
2. build a tarball + immediately publish it w/o testing that tarball.

What I advocate for is the following:
1. Build a tarball and store it as an artifact.
2. Have a test matrix that downloads that tarball, installs it and tests its contents (this is most likely a separate job or a collection of jobs).
3. Have a publish step that downloads the very same tarball that's been tested and uploads it to galaxy. This ensures that this step doesn't have a different build that's being tested.

artis3n commented:
@webknjaz that's a great point, and that's a flow I'll look into supporting. Feel free to file an issue if you'd like.

pabelanger commented:
> > FYI there's already a promising GitHub Action for publishing collections by @artis3n here: https://github.com/artis3n/ansible_galaxy_collection
> 
> That action also mixes up two stages that are meant to be separate: build + publish. I'm strongly against promoting such an approach.
> Ideally, it should be possible to test what's going to be published rather than just source. When build+publish are squashed there's a possibility that you'll be testing not exactly what the user will get. And this flow looks as follows:
> 
> 1. test the src.
> 2. build a tarball + immediately publish it w/o testing that tarball.
> 
> What I advocate for is the following:
> 
> 1. Build a tarball and store it as an artifact.
> 2. Have a test matrix that downloads that tarball, installs it and tests its contents (this is most likely a separate job or a collection of jobs).
> 3. Have a publish step that downloads the very same tarball that's been tested and uploads it to galaxy. This ensures that this step doesn't have a different build that's being tested.

This is what we do for testing network / security / content collections for ansible team.  It works very well, and also advocate for people to use.  One addition we are going to add, is an ephemeral galaxy server where we can store the artifact for testing, then promote to galaxy once code is merged.

webknjaz commented:
> One addition we are going to add, is an ephemeral galaxy server where we can store the artifact for testing, then promote to galaxy once code is merged.

Yeah, ansible/ansible now has "falaxy" for testing. Not sure if it can be used outside, though.



> @webknjaz that's a great point, and that's a flow I'll look into supporting. Feel free to file an issue if you'd like.

I don't have anything else to add there so you can copy what I wrote into an issue as you see fit. I'm currently implementing a similar workflow for a Python package distribution on GitHub Actions Workflows so it's slightly different packaging-wise but the you may get some high-level inspiration ideas from https://github.com/ansible/pylibssh/actions?query=workflow%3A%22%F0%9F%8F%97+%F0%9F%93%A6+%26+test+%26+publish%22 (some bits are still in progress tho).

ericsysmin commented:
> > FYI there's already a promising GitHub Action for publishing collections by @artis3n here: https://github.com/artis3n/ansible_galaxy_collection
> 
> That action also mixes up two stages that are meant to be separate: build + publish. I'm strongly against promoting such an approach.
> Ideally, it should be possible to test what's going to be published rather than just source. When build+publish are squashed there's a possibility that you'll be testing not exactly what the user will get. And this flow looks as follows:
> 
> 1. test the src.
> 2. build a tarball + immediately publish it w/o testing that tarball.
> 
> What I advocate for is the following:
> 
> 1. Build a tarball and store it as an artifact.
> 2. Have a test matrix that downloads that tarball, installs it and tests its contents (this is most likely a separate job or a collection of jobs).
> 3. Have a publish step that downloads the very same tarball that's been tested and uploads it to galaxy. This ensures that this step doesn't have a different build that's being tested.

Why do you have to be so correct....lol Can you make changes to this PR with that order?

webknjaz commented:
No, I don't have powers in this repo :)

webknjaz commented:
@ericsysmin you can find some inspiration on how to work with upload/download artifact @ https://github.com/ansible-community/collection_migration/blob/master/.github/workflows/collection-migration-tests.yml#L136-L188. Except that both now have `v2` released and the download one can fetch all the artifacts, not just one by name.

geerlingguy commented:
> 1. Build a tarball and store it as an artifact.
> 2. Have a test matrix that downloads that tarball, installs it and tests its contents (this is most likely a separate job or a collection of jobs).
> 3. Have a publish step that downloads the very same tarball that's been tested and uploads it to galaxy. This ensures that this step doesn't have a different build that's being tested.

I have a feeling we're talking about vastly different use cases at this point. For a certified content collection with Ansible modules that are organized by an enterprise vendor with a team of developers able to support it, having tooling in place for multiple layers of build, test, and deploy processes may be a worthy goal.

But I think both @ericsysmin and myself are coming at this from the approach of long-time Galaxy role authors.

We are used to having a lightweight CI system, a lightweight release system, and a lightweight package manager. We're coming from the angle of someone more used to systems like NPM, Packagist, or Rubygems, not from the angle of building and distributing RPMs, .debs, etc.

I think theres a major conflict in goals between the two use cases for collections. I just want there to be some easy way for me to maintain collections like I did roles, especially if [roles are going to vanish in Galaxy next-gen](https://github.com/ansible/galaxy_ng/issues/58)... because if it's an onerous process (like it is today), I know I'm not going to do it.

I don't want to have to tie two or three different backend systems/processes into each collection I maintain, and I'm probably more willing to bend on this topic than most people who have contributed roles in the past...

ericsysmin commented:
@geerlingguy it seems maybe a middle ground is that we do a build first, then do the install instead of doing this
```
before_script:
  - cd ../
  - mkdir -p ansible_collections/$COLLECTION_NAMESPACE
  - mv ansible-collection-$COLLECTION_NAME ansible_collections/$COLLECTION_NAMESPACE/$COLLECTION_NAME
  - cd ansible_collections/$COLLECTION_NAMESPACE/$COLLECTION_NAME
```

ericsysmin commented:
so really we'd be doing
```
before_script:
  - ansible-galaxy collection build
  - ansible-galaxy collection install my_namespace-my_collection-1.0.0.tar.gz

webknjaz commented:
@geerlingguy yeah, I was this approach a lot in the past and even now in the Python ecosystem, with publishing to PyPI and have it simplified like this in a lot of places.
One reason for this was that the public CIs didn't have a mature artifact and stage pipelining support. I feel like with GHA it became so much easier enabling virtually anyone can allow themselves a more sophisticated approach, especially if there'd be some pre-baked template that anybody could use.
Of course, some people may still want to opt out. And that's okay for them. OTOH we should show a more complete approach. Folks can then just remove the steps they dislike.

pabelanger commented:
> so really we'd be doing
> 
> ```
> before_script:
>   - ansible-galaxy collection build
>   - ansible-galaxy collection install my_namespace-my_collection-1.0.0.tar.gz
> ```

Yes, this is much better real world testing, then manually moving, creating the install manually.

ericsysmin commented:
SOoooo, trying this out, it may not be as easy as hoped, it also will require galaxy.yml being present at all times, so no version generation allowed, so people will have to always modify that file on release.

gundalow commented:
I see https://github.com/ansible-community/molecule/pull/2690 has been merged.

What do people think the way forward is here?

ssbarnea commented:
One thing that I find bit mind-blowing is that there is no command `ansible-galaxy collection ls`. No official way to list what collections are installed.

Also there is no `rm` either and that does not help much with automation or debugging.

felixfontein commented:
It's called `ansible-galaxy collection list` and has been working in devel since for some months now. :) (No `rm`/`uninstall` though.)

pabelanger commented:
> One thing that I find bit mind-blowing is that there is no command `ansible-galaxy collection ls`. No official way to list what collections are installed.
> 
> Also there is no `rm` either and that does not help much with automation or debugging.

Created https://github.com/ansible/ansible/issues/60292 a while back, happy to say code was added.

ssbarnea commented:
While I do have jobs testing deve, I am not brave enough to run devel locally so I didn't know about the list feature. Probably it should have being ported to latest stable.

Andersson007 commented:
@acozine SGTM, closing, thanks everyone!
If anyone is interested to proceed, feel free to re-open (or ping me)
